evap/staff/tools.py

```diff
@@ -188,3 +189,10 @@ def merge_users(main_user, other_user, preview=False):
     other_user.delete()
 
     return merged_user, errors, warnings
+
+def find_next_unreviewed_course(semester, excluded):
+    return semester.course_set.exclude(pk__in=excluded) \
+        .exclude(state='published') \
+        .annotate(num_unreviewed_textanswers=Count("contributions__textanswer_set", filter=Q(contributions__textanswer_set__state=TextAnswer.NOT_REVIEWED), distinct=True)) \
+        .filter(num_unreviewed_textanswers__gt=0) \
```

Comment:
did you try doing `.filter(contributions__textanswer_set__state=TextAnswer.NOT_REVIEWED)`? (not sure whether that works)

Reply:
test

Reply:
Yes, still on my try list. I thought that the feature of not cycling through two courses was more important than this experiment. At least, if I will say no, I want to give you a good explanation.

Reply:
With my small knowledge about Django and my failed attempt to implement this, I would say that it is not possible.

For reference: In the [docs](https://docs.djangoproject.com/en/2.0/topics/db/aggregation/#filtering-on-annotations) they show how to count all books rated higher than seven. So somehow it has to be done that the annotation filters out everything. I tried as alternative something like `Queryset("contributions__textanswer_set").filter(state=TextAnswer.NOT_REVIEWED).count()` but it does not connect correctly to the outer query.

Reply:
i tried this here and it seemed to just work:
```
# current solution
>>> Semester.objects.first().course_set\
    .annotate(num_unreviewed_textanswers=Count("contributions__textanswer_set", filter=Q(contributions__textanswer_set__state=TextAnswer.NOT_REVIEWED), distinct=True))\
    .filter(num_unreviewed_textanswers__gt=0).count()
43

# proposed solution
>>> Semester.objects.first().course_set\
    .filter(contributions__textanswer_set__state=TextAnswer.NOT_REVIEWED).distinct().count()
43

```

Reply:
Still, I am not sure where you want to apply this snippet. Your code will count the number of courses in a semester which have unreviewed text answers (That's for the filter toolbar in the semester overview).
The next course is the course with unreviewed text answers which has been ended most in the past. Do you want to propose a change how the next course is displayed or is this a replacement for the innermost query?

---------------------

code in git gives:
```sql
SELECT ..., COUNT(DISTINCT "evaluation_textanswer"."id")
FILTER (WHERE "evaluation_textanswer"."state" = NR) AS "num_unreviewed_textanswers"
FROM "evaluation_course"
LEFT OUTER JOIN "evaluation_contribution" ON ("evaluation_course"."id" = "evaluation_contribution"."course_id")
LEFT OUTER JOIN "evaluation_textanswer" ON ("evaluation_contribution"."id" = "evaluation_textanswer"."contribution_id")
WHERE ("evaluation_course"."semester_id" = 21 AND NOT ("evaluation_course"."state" = published))
GROUP BY "evaluation_course"."id"
ORDER BY "evaluation_course"."vote_end_date" ASC, "num_unreviewed_textanswers" DESC
```
your proposal gives:
```sql
SELECT ..., COUNT(DISTINCT "evaluation_textanswer"."id")
FILTER (WHERE "evaluation_textanswer"."state" = NR) AS "num_unreviewed_textanswers"
FROM "evaluation_course"
LEFT OUTER JOIN "evaluation_contribution" ON ("evaluation_course"."id" = "evaluation_contribution"."course_id")
LEFT OUTER JOIN "evaluation_textanswer" ON ("evaluation_contribution"."id" = "evaluation_textanswer"."contribution_id") WHERE "evaluation_course"."semester_id" = 21
GROUP BY "evaluation_course"."id"
HAVING COUNT(DISTINCT "evaluation_textanswer"."id")
FILTER (WHERE ("evaluation_textanswer"."state" = NR)) > 0
ORDER BY "evaluation_course"."name_de" ASC
```

Reply:
i proposed to replace the long annotate plus filter by a simple filter, sorry if that wasn't clear. the exclude and order need to stay of course.